### DISCUSSION TITLE
Clarity on Requirement

### DISCUSSION THREAD

**Discussion by dipak101505** [KathiraveluLab/Diomede#2] [General]
Hi @pradeeban ,

I'm Dipak Agarwal, an IIT Kgp graduate with over a year of experience working on cloud PACS. I'd like to discuss the project's problem statement to ensure clarity.

My understanding:

The project explores connecting a PACS system with an AI model (built in MATLAB) to process and triage DICOM files. The AI model would:

Analyze DICOM files: Generate heatmaps, outlines, and bounding boxes.
Send the response: Forward the processed data (including annotations and their probabilities) back to the PACS.
Confirmation and next steps:

Confirmation: Please confirm if my understanding of the problem statement is correct.
Coding challenge proposal: Could you provide guidance on what I can propose for the coding challenge related to this project? Should I write matlab code to analyze dcm file?
Best regards,

**Comment by pradeeban** [comment]
Hi @dipak101505,
Thanks for your interest.

The project is more of an "enabler" for machine learning, rather than building the AI models.

Your understanding is largely correct. While you can handle the "Analyze DICOM files" options that you have listed, those are not mandatory outcomes of the project. The project should just make such executions possible.

That means, if I have an algorithm (written in Matlab) that can run on DICOM images, can you provide an interface (a simple implementation) that can function as a wireframe to hook my ML functionality?

For example, https://github.com/Emory-HITI/Niffler is a DICOM framework in Python. Can we have a similar implementation in/for Matlab?

Does that explain better?

**Reply by dipak101505** [reply]
Hi @pradeeban ,

Thank you for your reply. I understand that accessing the MATLAB application directly through a port like in Python's Netfiller wasn't feasible.

However, I believe an alternative approach using a Docker image could be a viable solution. In this scenario, the standalone application could be packaged as a Docker image. This would allow us to execute the application using the docker run command, specifying required parameters like dicom_path and other relevant information, and obtain the desired output.

Here are some potential benefits of this approach:

Portability: Docker images are platform-agnostic, meaning they can run consistently across different operating systems without modification.
Isolation: Each containerized application runs in its own isolated environment, ensuring minimal conflicts with other applications or the underlying system.
Reproducibility: Docker images guarantee a consistent environment, making it easier to reproduce results across different systems.

would are your thoughts on this?

**Reply by pradeeban** [reply]
Hi @dipak101505,
Of course, in the past, we have run Docker instances on Niffler. So, surely we should support such execution from the proposed Diomede as well.

The details on how we did in Niffler are in the paper:
Pradeeban Kathiravelu, Puneet Sharma, Ashish Sharma, Imon Banerjee, Hari Trivedi, Saptarshi Purkayastha, Priyanshu Sinha, Alexandre Cadrin-Chenevert, Nabile Safdar, and Judy Wawira Gichoya. A DICOM Framework for Machine Learning Pipelines against Real-Time Radiology Images. Journal of Digital Imaging (JDI). August 2021. https://doi.org/10.1007/s10278-021-00491-w


Our goal in this project is to create a Matlab program that can:
* pull images on-demand from DICOM sources such as scanners and PACS.
* allow processing of images in Matlab directly.
* provide a DICOM listener to receive DICOM images from the DICOM sources.

While running Docker images is a good addition, that is not the goal of this project. This project is meant to be a Matlab implementation of a DICOM framework with the above functionalities. So, with time, this project can be a stand-alone project for Matlab enthusiasts to expand and improve its functionalities.

**Reply by dipak101505** [reply]
After reviewing the insightful research paper, the project requirements have become significantly clearer. However, implementing a real-time DICOM listener in MATLAB might pose a challenge due to the application not exposing ports for receiving requests. I'm eager to contribute to this project and would greatly appreciate your continued guidance. To best utilize my skills, could you advise where my contributions would be most valuable?
![nfiller](https://github.com/KathiraveluLab/Diomede/assets/26256895/bd0a6471-6108-4e23-859e-43c1972fef46)

**Reply by pradeeban** [reply]
aha, right. There could be missing or misinterpreted elements such as those. State those limitations in your proposal.

You could focus on the "Retrospective" on-demand extractions and then retrieving metadata and PNG images from those, in a unified manner, using Matlab then.

It is also a proper formulation as a project, rather than a bunch of code lying around. I think, we did a decent job with Niffler on that, although it still could be significantly formatted better (not part of this project. just thinking out loud).

**Reply by dipak101505** [reply]
**Addressing DICOM Listener and Proposing an Alternative**

I understand your initial aim was to implement a DICOM listener directly within the MATLAB application. While this approach offers real-time processing, it can be technically challenging due to limitations with native MATLAB functionalities for DICOM listening.

**Proposal: Adapter for Real-Time and Retrospective Processing**

To achieve both real-time and retrospective functionalities in a more feasible manner, I propose an adapter-based solution:

- Adapter as a Bridge: This adapter would act as a bridge between the PACS system and MATLAB application.

- Communication Channels: The adapter would connect to the PACS system via standard HTTP requests, retrieving DICOM data as needed. It would then utilize the [MATLAB Engine API](https://www.mathworks.com/help/matlab/matlab-engine-for-python.html) to seamlessly communicate with your MATLAB application, passing on the retrieved data for processing.

**Benefits**:

- Reusability: This structured approach promotes code reusability. The core MATLAB application logic can handle both real-time and retrospective data received through the adapter.

- Focus on Retrospective Extraction: We can focus on developing robust functions within the MATLAB application for retrospective on-demand data extraction (metadata and PNG images) and other functionalities(triaging).

**Project Structure and Prioritization**

This approach transforms your concept into a well-defined project with clear components and functionalities. It builds upon the successful aspects of your "Niffler" project, aiming for even better organization and maintainability.

**Next Steps:**

While we've explored potential solutions, further investigation is needed to refine the adapter design and implementation details.  We can then move on to building the retrospective processing functionalities within the MATLAB application.

By combining your initial vision with this alternative approach, we can achieve a robust and reusable solution for real-time **and** retrospective DICOM data processing in MATLAB.

**Reply by pradeeban** [reply]
@dipak101505 I like your alternative approach. Sounds like a great way to achieve this. Feel free to continue with your proposal draft.
